File: arrays_hashing/contains_duplicate/contains_duplicates_test.py
import pytest
from contains_duplicates import contains_duplicates


@pytest.mark.parametrize(
    "nums, expected",
    [
        ([1, 2, 3, 3], True),  # Case 1: Simple duplicate
        ([1, 2, 3, 4], False),  # Case 2: No duplicates
        ([], False),  # Case 3: Empty list
        ([1, 1, 1, 1], True),  # Case 4: All same elements
        ([-1, -2, -1], True),  # Case 5: Negative duplicates
        ([1], False),  # Case 6: Single element
    ],
)
def test_contains_duplicates(nums, expected):
    """
    Test the contains_duplicates function with various scenarios
    using pytest parameterization for clean, dry code.
    """
    assert contains_duplicates(nums) == expected

File: arrays_hashing/contains_duplicate/contains_duplicates_test.go
package containsduplicates

import "testing"

func TestContainsDuplicates(t *testing.T) {
	tests := []struct {
		name     string
		nums     []int
		expected bool
	}{
		{
			name:     "Has duplicates",
			nums:     []int{1, 2, 3, 3},
			expected: true,
		},
		{
			name:     "No duplicates",
			nums:     []int{1, 2, 3, 4},
			expected: false,
		},
		{
			name:     "Empty slice",
			nums:     []int{},
			expected: false,
		},
		{
			name:     "Negative numbers with duplicates",
			nums:     []int{-1, -2, -1},
			expected: true,
		},
		{
			name:     "All same numbers",
			nums:     []int{1, 1, 1, 1},
			expected: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ContainsDuplicates(tt.nums)
			if result != tt.expected {
				t.Errorf("ContainsDuplicates(%v) = %v; want %v", tt.nums, result, tt.expected)
			}
		})
	}
}
